naga/src/back/hlsl/help.rs

```diff
@@ -1031,6 +1051,208 @@ impl<W: Write> super::Writer<'_, W> {
+                    _ => continue,
+                };
+
+                let wrapped = WrappedUnaryOp {
+                    op,
+                    ty: (scalar, vector_size),
+                };
+
+                match (op, expr_ty.scalar()) {
+                    (
+                        crate::UnaryOperator::Negate,
+                        Some(crate::Scalar {
+                            kind: crate::ScalarKind::Sint,
+                            width: 4,
+                        }),
+                    ) => {
+                        if !self.wrapped.unary_op.insert(wrapped) {
+                            continue;
+                        }
+
+                        self.write_value_type(module, expr_ty)?;
+                        write!(self.out, " {NEG_FUNCTION}(")?;
+                        self.write_value_type(module, expr_ty)?;
+                        writeln!(self.out, " val) {{")?;
+
+                        let level = crate::back::Level(1);
+                        writeln!(self.out, "{level}return asint(~asuint(val) + 1u);",)?;
+                        writeln!(self.out, "}}")?;
+                        writeln!(self.out)?;
+                    }
+                    _ => {}
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    pub(super) fn write_wrapped_binary_ops(
+        &mut self,
+        module: &crate::Module,
+        func_ctx: &FunctionCtx,
+    ) -> BackendResult {
+        for (expr_handle, expression) in func_ctx.expressions.iter() {
+            if let crate::Expression::Binary { op, left, right } = *expression {
+                let expr_ty = func_ctx.resolve_type(expr_handle, &module.types);
+                let left_ty = func_ctx.resolve_type(left, &module.types);
+                let right_ty = func_ctx.resolve_type(right, &module.types);
+
+                match (
+                    op,
+                    func_ctx
+                        .resolve_type(expr_handle, &module.types)
+                        .scalar_kind(),
+                ) {
+                    (
+                        crate::BinaryOperator::Divide,
+                        Some(kind @ crate::ScalarKind::Sint | kind @ crate::ScalarKind::Uint),
+                    ) => {
+                        let left_wrapped_ty = match *func_ctx.resolve_type(left, &module.types) {
+                            crate::TypeInner::Scalar(scalar) => (scalar, None),
+                            crate::TypeInner::Vector { size, scalar } => (scalar, Some(size)),
+                            _ => unreachable!(),
+                        };
+                        let right_wrapped_ty = match *func_ctx.resolve_type(right, &module.types) {
+                            crate::TypeInner::Scalar(scalar) => (scalar, None),
+                            crate::TypeInner::Vector { size, scalar } => (scalar, Some(size)),
+                            _ => unreachable!(),
+                        };
```

Comment:
Aren't these `resolve_type` calls exactly `left` and `right`, already computed above?

Reply:
yes again